### PR TITLE
WL-3933: Get book thumbnail images from Google Books

### PR DESCRIPTION
**Pull request by nicholaswilson100**

### CODE REVIEW

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/BaseCitationService.java [hunk 1]
```diff
@@ -2205,6 +2200,24 @@ else if(this.m_urls.containsKey(urlid))
...
 		}
 
+		@Override
+		public String getThumbnailURL() {
+			String imgURL = null;
+			try {
+				URL url = new URL(GOOGLE_BOOKS_SEARCH_API + "q=isbn:" + getCitationProperty("isnIdentifier"));
+				BufferedReader input = new BufferedReader(new InputStreamReader(url.openStream(), "UTF-8"));
+				JSONObject obj = new JSONObject(IOUtils.toString(input));
+				imgURL = (String) ((JSONObject)obj.getJSONArray("items").get(0)).getJSONObject("volumeInfo").getJSONObject("imageLinks").get("smallThumbnail");
```
Comment (buckett) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/BaseCitationService.java:2210]: What happens when there isn't an items array in the response?

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/BaseCitationService.java [hunk 2]
```diff
@@ -2205,6 +2200,24 @@ else if(this.m_urls.containsKey(urlid))
 			}
 		}
 
+		@Override
+		public String getThumbnailURL() {
+			String imgURL = null;
+			try {
+				URL url = new URL(GOOGLE_BOOKS_SEARCH_API + "q=isbn:" + getCitationProperty("isnIdentifier"));
+				BufferedReader input = new BufferedReader(new InputStreamReader(url.openStream(), "UTF-8"));
```
Comment (buckett) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/BaseCitationService.java:2208]: What's the timeout on the request? Can problems accessing the google API mean that the reading list fails to render?

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/BaseCitationService.java (hunk 1 above)
Comment (nicholaswilson100) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/BaseCitationService.java:2210]: We show the placeholder book cover (another Jira)

**Comment by buckett** [comment]
Generally, is there a reason todo this serverside, rather than clientside?

**Comment by buckett** [comment]
Didn't mean to close, clicked wrong.

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/BaseCitationService.java (hunk 1 above)
Comment (buckett) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/BaseCitationService.java:2210]: Won't it throw a NPE?

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/BaseCitationService.java (hunk 2 above)
Comment (nicholaswilson100) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/BaseCitationService.java:2208]: Good question.. what do you think? Shall I put some timeout on a request of a second or two maximum?

**Comment by nicholaswilson100** [comment]
There wasn't a particularly strong reason to do it server-side.  I went that way because it's less code and this Java method is more easily testable than something in Javascript, ... are you thinking async ajax might have been better so we aren't waiting on Google Books?

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/BaseCitationService.java (hunk 1 above)
Comment (nicholaswilson100) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/BaseCitationService.java:2210]: Ah I see what you mean.. yes I'll change that

**Comment by buckett** [comment]
If you've got 50 books on a page you've got to wait for 50 calls to google to complete and they are serial, due to how the rendering of the page is done. If any one of them hangs the whole reading list won't be displayed until it times out.

The failure just seems much worse when you do it in Java and as it's unauthenticated doing it in the browser doesn't lose you much.

**Comment by nicholaswilson100** [comment]
Yup ok good point, I'll move it client-side.

**Comment by buckett** [comment]
Would be good to isolate this in a jQuery plugin so it's pretty self contained.

**Comment by nicholaswilson100** [comment]
I'm just going to do a fresh pull request rather than reverting, squishing, committing. Easier and cleaner.

**Pull request closed without merging**